WordPressKit/Authenticator.swift

```diff
@@ -0,0 +1,231 @@
+    }
+}
+
+// MARK: Private helpers
+private extension CookieNonceAuthenticator {
+
+    func startLoginSequence(manager: SessionManager) {
+        DDLogInfo("Starting Cookie+Nonce login sequence for \(loginURL)")
+        guard let nonceRetrievalURL = nonceRetrievalMethod.buildURL(base: adminURL) else {
+            return invalidateLoginSequence(error: .invalidNewPostURL)
+        }
+        let request = authenticatedRequest(redirectURL: nonceRetrievalURL)
+        manager.request(request)
+            .validate()
+            .responseString { [weak self] (response) in
+                guard let self = self else {
+                    return
+                }
+                switch response.result {
+                case .failure(let error):
+                    self.invalidateLoginSequence(error: .postLoginFailed(error))
+                case .success(let page):
+                    let redirectedTo = response.response?.url?.absoluteString ?? "nil"
+                    DDLogInfo("Posted Login to \(self.loginURL), redirected to \(redirectedTo)")
+                    guard let nonce = self.nonceRetrievalMethod.retrieveNonce(from: page) else {
+                        return self.invalidateLoginSequence(error: .missingNonce)
+                    }
+                    self.nonce = Secret(nonce)
+                    self.successfulLoginSequence()
+                }
+        }
+    }
+
+    func successfulLoginSequence() {
+        DDLogInfo("Completed Cookie+Nonce login sequence for \(loginURL)")
+        completeRequests(true)
+    }
+
+    func invalidateLoginSequence(error: Error) {
+        canRetry = false
+        DDLogError("Aborting Cookie+Nonce login sequence for \(loginURL)")
+        completeRequests(false)
+        isAuthenticating = false
+    }
+
+    func completeRequests(_ shouldRetry: Bool) {
+        requestsToRetry.forEach { (completion) in
+            completion(shouldRetry, 0.0)
+        }
+        requestsToRetry.removeAll()
+    }
+
+    func authenticatedRequest(redirectURL: URL) -> URLRequest {
+        var request = URLRequest(url: loginURL)
+
+        request.httpMethod = "POST"
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+
+        var parameters = [URLQueryItem]()
+        parameters.append(URLQueryItem(name: "log", value: username))
+        parameters.append(URLQueryItem(name: "pwd", value: password.secretValue))
+        parameters.append(URLQueryItem(name: "rememberme", value: "true"))
+        parameters.append(URLQueryItem(name: "redirect_to", value: redirectURL.absoluteString))
+        var components = URLComponents()
+        components.queryItems = parameters
+        request.httpBody = components.percentEncodedQuery?.data(using: .utf8)
+        return request
+    }
+
+}
+
+public extension CookieNonceAuthenticator {
+    static func cookieNonce(username: String, password: String, loginURL: URL, adminURL: URL, version: String) -> Authenticator {
```

Comment:
> static func cookieNonce(...)

This method doesn't look necessary since CookieNonceAuthenticator is public. One could simply use the initializer instead.

But, if the intention was to keep CookieNonceAuthenticator private and provide a kind of factory method for `Authenticator`(which I think would be better) then it would make sense to keep it but make it an extension for Authenticator instead:

```
public extension Authenticator { 
     static func cookieNonce(username: String, password: String, loginURL: URL, adminURL: URL, version: String) -> Authenticator {
....
```

Reply:
Same is true for:

```
public extension TokenAuthenticator {
    static func token(_ token: String, shouldAuthenticate: RequestAuthenticationValidator? = nil) -> Authenticator {
        return TokenAuthenticator(token: token, shouldAuthenticate: shouldAuthenticate)
    }

}

```

Reply:
I agree that they're not very useful now. Those were initially created to be able to use the authenticators more like enums, so you would be able to do something like:

```swift
let authToken = "secret"
let api = WordPressOrgRestApi(
    url: baseURL,
    authenticator: .token(authToken)
)
```

However, since I ended up with [a convenience initializer](https://github.com/wordpress-mobile/WordPress-iOS/blob/92d7354c7342c41caaec4c8d4488161f0a7b3f64/WordPress/Classes/Networking/WordPressOrgRestApi%2BWordPress.swift#L37) that takes a `Blog`, these aren't being used in that way and can be removed.

Reply:
For the record, I needed to change the code from being extension of Authenticator to the specific classes because of an error that the Swift compiler was returning. 
I assume this must have been some recent change on the Swift compiler, and I will remove the methods in favor of the blog init helpers.